### PR TITLE
[RFC] WT-8682: Remove session from some function calls

### PR DESCRIPTION
**Pull request by korteland**
This is a draft pull request for one possible approach to removing the session pointer from a large number of function calls. The eventual goal of this work is to make unit testing feasible. Please feel free to remove yourself as a reviewer if you don't want to look at it; I just want some experienced eyes on this before I forge ahead.

A few points:

The key changes are in `session.h`. The main risks here are the "union of structs" thing, which would only have to last until the refactoring work is done, and the fact that there are now two identical `WT_CONNECTION` pointers in `__wt_session_impl`. One is added to the new "metadata" struct (name is a WIP, suggestions welcome) and the existing one in `iface`. I didn't want to include the entire `iface` in the `metadata` struct, but opinions on this are welcome.

Secondly, the overall thing I'm trying to show is what it would look like for this work to continue. It should be visible at this point how the metadata struct would get "hoisted up" along the call stack as this work progresses, eventually winding up at a hopefully small number of functions that genuinely need the entire session. This might include further breaking down the session, or removing parts of the metadata struct - I'll cross that bridge when I come to it.

Finally - thanks for any time you spend on this. There's no pressure to merge it, it's an experiment and I've got about 2.5 days into it. Saying the entire approach is wrong and I should do something else is an outcome I'm happy with.

### CODE REVIEW

File: src/include/session.h [hunk 1]
```diff
@@ -52,19 +53,68 @@ typedef TAILQ_HEAD(__wt_cursor_list, __wt_cursor) WT_CURSOR_LIST;
...
+            WT_ITEM err; /* Error buffer */
+
+            WT_SESSION_STATS stats;
+
+            /* Sessions have an associated statistics bucket based on its ID. */
+            u_int stat_bucket; /* Statistics bucket offset */
+        };
+
+        struct __wt_session_metadata metadata;
+    };
```
Comment (korteland) [inline src/include/session.h:112]: Just FYI, the weird "union of structs" thing would go away once this is reasonably complete. It's just a hack to let existing code continue working while stuff gets moved out of `__wt_session_impl`.

File: src/include/time_inline.h [hunk 2]
```diff
@@ -48,20 +48,20 @@ __wt_rdtsc(void)
  *     highest value seen so far.
  */
 static inline void
-__time_check_monotonic(WT_SESSION_IMPL *session, struct timespec *tsp)
+__time_check_monotonic(WT_SESSION_METADATA *meta, struct timespec *tsp)
```
Comment (korteland) [inline src/include/time_inline.h:51]: This was where it all started - I picked this function and started thinking about "what would it take to remove the session passed in here?"

**Comment by korteland** [comment]
It passes the ctest suite and python test suite, haven't tried anything else yet.

**Comment by keithbostic** [comment]
I dislike this change -- as noted in the PR, the name `metadata` is essentially meaningless in this context, and I don't understand why having N buckets of undifferentiated stuff is better than having one bucket of undifferentiated stuff. I believe there might be places where we could abstract out chunks of information from the `WT_SESSION_IMPL` handle (for example, we already do so with the block manager chunk and the transaction state). In other words, I'm not opposed to information abstraction as a goal, but my prediction is this change will require lots of work for little benefit.

That said -- I agree that I'm not giving this change a fair hearing and I'm probably too invested in the current structure to weigh the benefits with an open mind. I'll take myself off the review group, other folks less invested in the current structure should make this decision.

**Comment by korteland** [comment]
That's fair enough. Thanks for the review, I'll close this and keep exploring. While we're here, do you have any ideas on making the code a bit more unit-testable? I was focused on removing the giant session structure from the arguments for a bunch of functions, but there could be better approaches that I haven't seen.

**Comment by keithbostic** [comment]
> That's fair enough. Thanks for the review, I'll close this and keep exploring.

I was serious, I'm too invested in the existing structure -- this might be welcomed by other engineers in the group.

> While we're here, do you have any ideas on making the code a bit more unit-testable? I was focused on removing the giant session structure from the arguments for a bunch of functions, but there could be better approaches that I haven't seen.

Why does requiring a session structure impede unit testing? No matter what information the target function requires, you have to dummy it up for unit testing, right? Why is it unreasonable to dummy up a large structure vs. a smaller one?

**Comment by korteland** [comment]
> I was serious, I'm too invested in the existing structure -- this might be welcomed by other engineers in the group.

Perhaps it would be welcome, but the points you raised above (around the work/benefit tradeoff) are perfectly valid; and people can still make comments on these changes.

> Why does requiring a session structure impede unit testing? No matter what information the target function requires, you have to dummy it up for unit testing, right? Why is it unreasonable to dummy up a large structure vs. a smaller one?

It's not so much the size of the structure but the complexity of it. If I'm targeting (say) `__wt_sweep_destroy` (or whatever, this is just an example) it's hard to reason about how it's using `session`. Do I need to provide "realistic" values of the `cursors` field? What about `active`? etc. This answer will be different for every function call, as is what constitutes "realistic" example data.

This `__wt_sweep_destroy` example would probably end up taking this new metadata struct and some form of sweep-specific structure. The metadata structure added here is simple enough that (in my opinion, at least) it's easy to reason about and provide a "ready made" copy in the tests, but still has the fields needed by a large subset of functions that only need a session so they can pass it to `__eventv`. This would let us worry solely about sweep-specific data in this example.

**Comment by sueloverso** [comment]
I'm definitely in the same boat as @keithbostic. Right now I know, no matter what code I'm in that I have a session available to me. From that session I can get to the global connection, I can get to my dhande or btree. I don't have to think about it at all. If all functions now take some subset, or sub-structure of a session, I have to think about what function takes what piece and what to do if I need more than that one piece.

I suspect that trying to make any single function self-contained and minimal for some sort of unit testing is going to be at odds with engineering and developer ease of use of the code and being able to reason about it. So it doesn't seem that the change is for developer benefit and is likely to add to developer time/cost.

**Comment by ddanderson** [comment]
I admire the goal of getting code that can be unit tested.  But I think this is going to cause a different kind of maintenance burden, where we need to "change the plumbing" more often as code evolves.  Have we tried the less invasive route in constructing a "dummy session" to unit test `time_check_monotonic` or `__wt_sweep_destroy`?  It seems to me that we could have a standard way to create such a "dummy session", and as that constructor gets more sophisticated, more functions will be available for unit testing.

**Comment by korteland** [comment]
Thanks for the feedback everyone - I appreciate it, positive or negative.

Sue - Point taken about expecting to always have the session available. I'm relatively new to the code so it wasn't a problem I had personally, and this sort of comment is useful to me.

Don - thanks for the suggestion. I had a look at it, but didn't end up writing any code. It seems difficult to provide a session that's in a reasonable state for anything that might want to use it, but if we don't want to go down the "remove session from function calls" path it might be the only option.

**Comment by keitharnoldsmith** [comment]
@korteland : Apologies for showing up late here -- after you've closed the PR, apparently. As you might have guessed from the interest in this diff/RFC, this came up at a Tora meeting earlier today and so we've all stopped by to offer our thoughts.  I hope it doesn't seem like we're piling on :-)

I agree with others who point out that we often assume we can access state in `WT_SESSION_IMPL` from wherever we want.  But at the same time, there are things in that structure that clearly *shouldn't* be getting used in arbitrary places.  If the history store code is manipulating `session->compact` we have a problem.  So there are some parts of the session we should only need to provide for testing specific pieces of code. 

This suggests to me a variation on @ddanderson's proposal.  Rather than working to develop an increasingly general "dummy session" that can be used anywhere, we should have different flavors of dummy session that we use in different places.  There will be some general stuff that we always want to have available.  Then each module will have additional fields it needs to include in its version of the dummy session.  For example, a quick grep through the block manager code shows that it never uses `session->dhandle` although that field is commonly used in lots of other places, the block manager is also unlikely to need the cursor cache and lots of other things in `WT_SESSION_IMPL`.

So we'd wind up with a collection of session state that we need for testing the block manager, and a different way of setting up a session for testing the logging system, or metadata tracking, etc.  This seems in keeping with the goal of increasing the modularity of the code (or of clarifying the existing boundaries between modules).  It also suggests that we might wind up more explicitly grouping related content in the session, whether that is by creating explicit sub-structures or just grouping things appropriately in the header (which we largely do already, I think).

I don't know if this is feasible, but I suspect that part of the complexity here is that we don't have any of this yet.  If there were already 100 unit tests for the block manager and I wanted to add a few more, I wouldn't have to figure out what to put in the session for my new tests; I'd just use the same dummy session the other 100 tests use.

**Comment by korteland** [comment]
No worries - pile away :-) If there's concern about the pull request being closed this might be better as an email thread, I just closed it to indicate I'm no longer considering merging code along these lines.

I like your suggestion as a middle ground. I agree that we already group things in the header file, and that sub-structures might provide a stronger level of isolation, although if the information is all still there I'm not sure it solves the problem of it being a complex structure that's difficult to reason about.

I agree having a set of different dummy sessions appropriate for different subsystems would get us at least some way towards unit-testing WiredTiger. I'm not sure how we'd go longer-term since some subsystems touch a lot more stuff than others, and some subsystems interact with others, but only in a minor way (for example, dhandles need to know about a connection's sweep interval and idle time). But maybe we should just cross that bridge when we come to it?

**Comment by jeremythorp** [comment]
Thanks @korteland , @keithbostic , @sueloverso , @ddanderson , @keitharnoldsmith for a great ongoing discussion!

Creating 'dummy' data structures is potentially very useful for testing code in isolation, but is potentially challenging if a data structure is large/complex because being able to confirm that a 'dummy' structure is correct for a particular test becomes a challenge. For example,`__wt_session_impl` has over 50 members, and `__wt_connection_impl` has over 200. I'm skeptical of complex mocks for similar reasons - if a mock is complex, how do you know if it is correct (if the mock has a bug, you're in for a world of pain). One the other hand, simple mocks can be very useful. 

I've generally found engineers find it easier to work on code when they only have to reason about (and remember the details of) a small(ish) amount of code/data/info at any one instant in time. When discussing code complexity, Steve McConnell (in [Code Complete](https://en.wikipedia.org/wiki/Code_Complete), 2nd ed) argues that "_Moving part of a routine into another routine doesn’t reduce the overall complexity of the program; it just moves the decision points around. But it reduces the amount of complexity you have to deal with at any one time. Since the important goal is to minimize the number of items you have to juggle mentally, reducing the complexity of a given routine is worthwhile._"

The [The Art of Unix Programming](http://www.catb.org/esr/writings/taoup/html/index.html) suggests [that](http://www.catb.org/~esr/writings/taoup/html/ch04s01.html) "_small differences in the maximum chunk size a programmer can hold in short-term memory have a large multiplicative effect on the programmer's efficiency. This might be a major contributor to the order-of-magnitude (or larger) variations in effectiveness observed by Fred Brooks and others._” and asks “_Do your data structures have more than seven members?_” (ie implying that more than 7 is not ideal)

While this particular PR has not been (so far) been successful in improving WT, the fact that 61 files would have to be changed to make the change that Will suggested is in itself, to me at least, a sign that the structure of the code and data structure use is potentially more complex than is ideal. The PR has also been a trigger for a great discussion :-).

**Comment by sueloverso** [comment]
@korteland can I suggest that this discussion be copied-and-pasted back into the ticket so that we have a JIRA history that is searchable, etc?

**Comment by korteland** [comment]
Yep, good call. Thanks.

**Comment by jeremythorp** [comment]
@sueloverso , that's a great idea!

**Pull request closed without merging**